### DISCUSSION TITLE
Hello, I made a whole test case by importing Driver. Now I'm realizing I need methods from the BaseCase like .switch_to_window... How to resolve?

### DISCUSSION THREAD

**Discussion by disposablecarcass** [seleniumbase/SeleniumBase#2347] [Q&A] [answered]
Any suggestions on how to fix this? I didn't write anything in a class. I usually just write a function like driver.get("url"). Then driver.click(). etc. 

I wrote my whole case with Driver: 

import seleniumbase as Driver
driver = Driver(uc=True)

driver.get("url")
driver.click(element)
etc.

But when I got to a point where I had a window pop up, I realized that there is no driver.switch_to_window() method. Is there a workaround for this? 

I noticed all of the examples are in classes that call themselves? Is there any way around this? I really just want to be able to do driver.switch_to_window()

Thank you

**Comment by mdmintz** [comment, marked as answer]
Look for any of the tests in the [SeleniumBase/examples/](https://github.com/seleniumbase/SeleniumBase/tree/master/examples) folder that start with `raw_`, as those are pure Python tests. (The tests start start with `test_` or end with `_test.py` are tests that are meant to be run with `pytest`, and those tests have that special class structure that you've noticed.) Here are those `raw_` tests:

<img width="334" alt="Screenshot 2023-12-07 at 7 49 30 PM" src="https://github.com/seleniumbase/SeleniumBase/assets/6788579/d6f56ccf-4e8b-4628-9c50-4f42ea7caa36"><br />

For example, here's a script that imports `Driver`:

```python
"""Driver() test. Runs with "python". (pytest not needed)."""
from seleniumbase import Driver

driver = Driver(browser="chrome", headless=False)
try:
    driver.open("seleniumbase.io/apps/calculator")
    driver.click('[id="4"]')
    driver.click('[id="2"]')
    driver.assert_text("42", "#output")
    driver.highlight("#output", loops=6)
finally:
    driver.quit()

driver = Driver()
try:
    driver.open("seleniumbase.github.io/demo_page")
    driver.highlight("h2")
    driver.type("#myTextInput", "Automation")
    driver.click("#checkBox1")
    driver.highlight("img", loops=6)
finally:
    driver.quit()
```

Here's a script that uses `SB()`: (Note that `sb` can use SeleniumBase BaseCase methods.)

```python
"""SB Manager using UC Mode for evading bot-detection."""
from seleniumbase import SB

with SB(uc=True) as sb:
    sb.driver.uc_open_with_tab("https://nowsecure.nl/#relax")
    sb.sleep(1.2)
    if not sb.is_text_visible("OH YEAH, you passed!", "h1"):
        sb.get_new_driver(undetectable=True)
        sb.driver.uc_open_with_reconnect(
            "https://nowsecure.nl/#relax", reconnect_time=3
        )
        sb.sleep(1.2)
    if not sb.is_text_visible("OH YEAH, you passed!", "h1"):
        if sb.is_element_visible('iframe[src*="challenge"]'):
            with sb.frame_switch('iframe[src*="challenge"]'):
                sb.click("span.mark")
                sb.sleep(2)
    sb.activate_demo_mode()
    sb.assert_text("OH YEAH, you passed!", "h1", timeout=3)
```

**Reply by disposablecarcass** [reply]
Thanks, but I don't seem to have access to the driver.switch_to_window method? In short I'm trying to use undetectable chrome driver with proxies. I'm done most of the script, but it switches windows at the end and I can't access that window. I was inspired to start this due to your recent video (well done) and some issues I've run into with undetectable-chromedriver project alone. 

I'd prefer to stick to importing Driver as that's closer to what I'm familiar with. With that said, it seems like I need BaseCase to use the .switch_to_window method? Driver doesn't appear to have any way to switch window on its own. 

It looks like I'd need to start writing everything in a class with BaseCase? Is that correct? Is there a workaround?

**Reply by mdmintz** [reply]
The `Driver()` format doesn't have all the methods that the `SB()` format has.

There are example scripts that use `SB()`: (Note that `sb` can call SeleniumBase BaseCase methods.)

Here's one:

```python
"""SB Manager using UC Mode for evading bot-detection."""
from seleniumbase import SB

with SB(uc=True) as sb:
    sb.driver.uc_open_with_tab("https://nowsecure.nl/#relax")
    sb.sleep(1.2)
    if not sb.is_text_visible("OH YEAH, you passed!", "h1"):
        sb.get_new_driver(undetectable=True)
        sb.driver.uc_open_with_reconnect(
            "https://nowsecure.nl/#relax", reconnect_time=3
        )
        sb.sleep(1.2)
    if not sb.is_text_visible("OH YEAH, you passed!", "h1"):
        if sb.is_element_visible('iframe[src*="challenge"]'):
            with sb.frame_switch('iframe[src*="challenge"]'):
                sb.click("span.mark")
                sb.sleep(2)
    sb.activate_demo_mode()
    sb.assert_text("OH YEAH, you passed!", "h1", timeout=3)
```

Here's another:

```python
from seleniumbase import SB

def open_the_turnstile_page(sb):
    sb.driver.uc_open_with_reconnect(
        "https://seleniumbase.io/apps/turnstile", reconnect_time=2.27,
    )

def click_turnstile_and_verify(sb):
    sb.driver.uc_switch_to_frame("iframe")
    sb.driver.uc_click("span.mark")
    sb.assert_element("img#captcha-success", timeout=3.33)

with SB(uc=True, test=True) as sb:
    open_the_turnstile_page(sb)
    try:
        click_turnstile_and_verify(sb)
    except Exception:
        open_the_turnstile_page(sb)
        click_turnstile_and_verify(sb)
    sb.set_messenger_theme(location="top_left")
    sb.post_message("Selenium wasn't detected!", duration=3)
```

For a proxy, there's the `proxy` arg. Formats: `host:port` or `user:pass@host:port`.

Switching windows can get you detected, unless you use the built-in `open`/`get` methods of UC Mode that open a new URL in a new tab and then immediately close the existing one to evade detection.

**Reply by disposablecarcass** [reply]
Okay thanks for the detailed reply. Perhaps I'll explore this further in the future.

**Reply by disposablecarcass** [reply]
Hey,  I realize this is a bit of a long shot and directly related to selenium base, but figured I'd at least ask. I accidentally broke my project the other day when playing around with some updates. I've worked on this project for a couple years now and I think I may have completely broken it. I've spent all day trying to fix it, but no luck. I saw you have helped others with undetected-chromedriver issues. Here is what I've done:

https://github.com/ultrafunkamsterdam/undetected-chromedriver/discussions/1695

Long story short, I updated my browser and changed a few of the package versions in an attempt to stay up to date and very much broke it. I recorded my setting ahead of time, but when reverting back I still can not get it to work without getting this error:
```
 from selenium.webdriver.chromium.options import ChromiumOptions as _ChromiumOptions
ModuleNotFoundError: No module named 'selenium.webdriver.chromium'
```
It was previously working on Chrome Version 112, Selenium version 3.9.0 and undetected-chromedriver version 3.4.6. I've since reverted to these settings but it still gives me the same error. 
Upon inspecting the package I realized that Chromium is not the Selenium folder for this version yet undetected-chrome driver is looking for it. I'm so confused as to what happened. Any idea what configuration of package versions will work for Chrome version 112?
I've spent the last day trying to fix this. Any insight would mean the world to me. Thank you

**Reply by mdmintz** [reply]
Try the known SeleniumBase UC Mode examples:
* https://github.com/seleniumbase/SeleniumBase/blob/master/examples/verify_undetected.py
* https://github.com/seleniumbase/SeleniumBase/blob/master/examples/uc_cdp_events.py
* https://github.com/seleniumbase/SeleniumBase/blob/master/examples/raw_uc_mode.py

And here are some newer examples where clicking is required:
* https://github.com/seleniumbase/SeleniumBase/blob/master/examples/raw_turnstile.py
* https://github.com/seleniumbase/SeleniumBase/blob/master/examples/raw_form_turnstile.py

**Reply by disposablecarcass** [reply]
Thanks, but my question is specific to the undetected-chromedriver package. Just trying to get the selenium/uc versions necessary to get rid of that error in chrome versoin 112. 

No clue why it stopped working, but maybe ultrafunk can explain if he sees my issue here: https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/1697

Anyway, thanks for your time.
